### ISSUE TITLE
Spaghetti Code

### ISSUE THREAD

**Issue by elbrujohalcon** [inaka/elvis#30]
##### Rule

> Don't write spaghetti code

![FSM](https://encrypted-tbn2.gstatic.com/images?q=tbn:ANd9GcS3GWtx0mzeTa-vhh8t4d5Tb9sdvfP9meBF5mwkz2CaZM96m_uFGVYTHHfF)

**Comment by jfacorro** [comment]
@igaray According to the rule's reasoning:

> Reasoning: Spaghetti code is harder to read, understand and edit. The function callgraph for your program should strive to be a directed acyclic graph.

A possible implementation would involve figuring out the call graph between functions in the whole project and detect where a cycle exists (I think recursive calls should not be considered a cycle though). This could involve a lot of computation, but even so, what should `elvis` report as the result of the code analysis? An entire module? A single function? A subset of expressions? 

We should decide what qualifies as spaghetti code, as far as I can see these are possible criteria:

1. **Coupling between modules**: this would compute a coupling degree between each of the modules in the project and report those with a degree higher than a specified level. 
2. **Number of cycles in the call graph**: If there are more than `n` cycles in the call graph then report the cycles.
3. **Depth of function calls**: Given a function defined in the project, if there are no cycles, compute the maximum depth of function calls.

These are not necessarily the best criteria and more can be thought of, but before we can implement the rule we need a more formal specification.

**Comment by jfacorro** [comment]
I will close this issue as "wontfix" since there is no clear and single way of defining spaghetti code. 

I think implementing a rule to measure cyclomatic complexity (mentioned in #24 ) is very related to what is described here. We can keep adding more and more rules that aim to reduce the complexity but a single rule that achieves this is "impossible" to achieve.
